src/mesh_tools.rs

```diff
@@ -150,49 +150,36 @@ pub fn separate_isolated_meshes(geometry: &Geometry) -> Vec<Geometry> {
     let mut available_face_indices: HashSet<u32> = face_to_face.keys().copied().collect();
     let mut patches: Vec<Geometry> = Vec::new();
 
-    while let Some(first_face_index) = available_face_indices.iter().next() {
-        let connected_indices = crawl_faces(*first_face_index, &face_to_face);
+    while let Some(start_face_index) = available_face_indices.iter().next() {
+        let mut index_stack = vec![*start_face_index];
+        let mut connected_face_indices = HashSet::new();
```

Comment:
Unless I misunderstand something, I'm pretty positive we can assume this is gonna have the same length as the original geometry. We'll also need `connected_face_indices.shrink_to_fit();` at the end as in previous iteration of the code.